src/Lykke.Service.EasyBuy.DomainServices/Timers/Timer.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Log;
+using Lykke.Common.Log;
+
+namespace Lykke.Service.EasyBuy.DomainServices.Timers
+{
+    public abstract class Timer
+    {   
+        protected ILog Log;
+
+        private bool _started;
+        private Task _task;
+        private CancellationTokenSource _cancellationTokenSource;
+
+        public void Start()
+        {
+            if (_started)
+                return;
+
+            _started = true;
+
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            _task = ExecuteAsync(_cancellationTokenSource.Token);
+        }
+
+        public void Stop()
+        {
+            if (!_started)
+                return;
+
+            _cancellationTokenSource?.Cancel();
+            _cancellationTokenSource?.Dispose();
+
+            _task?.GetAwaiter().GetResult();
+            _task?.Dispose();
+
+            _task = null;
+            _cancellationTokenSource = null;
+
+            OnStop();
+
+            _started = false;
+        }
+
+        protected virtual void OnStop()
+        {
+        }
+
+        protected abstract Task OnExecuteAsync(CancellationToken cancellation);
+
+        protected abstract Task<TimeSpan> GetDelayAsync();
+
+        private async Task ExecuteAsync(CancellationToken cancellation)
+        {
+            while (!cancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    await OnExecuteAsync(cancellation);
+                }
+                catch (Exception exception)
+                {
+                    Log.Error(exception);
```

Comment:
Could lead to NullReferenceException if Log is not initialized by inherited type. This could be resolved by these options:
1) check for Log == null in Start methid
2) Add explicit c-tor with ILog/ILogFactory parameter

Reply:
The log should be created in a specific instance, not in base class. I agree to add an additional check to prevent NRE.